File: tests/r_parity/51_newey.py
"""StatsPAI Newey-West HAC parity (Python side) -- Module 51."""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "51_newey"

def make_data(T=200, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x = rng.normal(0, 1, T)
    e = rng.normal(0, 1, T)
    y = np.zeros(T)
    for t in range(1, T):
        y[t] = 0.5 * y[t-1] + 0.3 * x[t] + e[t]
    return pd.DataFrame({"y": y, "x": x, "t": np.arange(T)})

def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.regress("y ~ x", df, robust="hac", lags=4)
    rows = []
    for nm, lab in [("Intercept", "intercept"), ("x", "x")]:
        if nm in res.params.index:
            rows.append(ParityRecord(MODULE, "py", f"beta_{lab}",
                estimate=float(res.params[nm]),
                se=float(res.std_errors[nm]),
                n=int(len(df))))
    write_results(MODULE, "py", rows, extra={"vcov": "HAC", "lags": 4})

if __name__ == "__main__":
    main()

File: tests/r_parity/51_newey.R
# StatsPAI Newey-West HAC parity (R side) -- Module 51.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(sandwich); library(lmtest) })

MODULE <- "51_newey"
df <- read_csv_strict(MODULE)
fit <- lm(y ~ x, data = df)
# NeweyWest with lag=4 (no prewhiten, no adjust to match Stata's `newey`)
V <- NeweyWest(fit, lag = 4, prewhite = FALSE, adjust = FALSE)
co <- coef(fit); se <- sqrt(diag(V))

rows <- list(
  parity_row(MODULE, "beta_intercept", estimate = unname(co["(Intercept)"]), se = unname(se["(Intercept)"]), n = nrow(df)),
  parity_row(MODULE, "beta_x",         estimate = unname(co["x"]),           se = unname(se["x"]),           n = nrow(df))
)
write_results(MODULE, rows, extra = list(vcov = "NeweyWest", lag = 4, package = "sandwich::NeweyWest"))
